### PR TITLE
port truncator to use visitor

### LINKED ISSUE #382: port truncator to be a visitor
Once https://github.com/rust-lang/chalk/pull/373 lands, we should port the truncator to use it.

cc @Areredify

### PR DESCRIPTION
**Pull request by Areredify**
closes #382 
it causes some (potentially useful?) code to become dead, should I remove it?

### CODE REVIEW

File: chalk-solve/src/infer.rs [hunk 1]
```diff
@@ -80,6 +80,7 @@ impl<I: Interner> InferenceTable<I> {
     }
 
     /// Current maximum universe -- one that can see all existing names.
+    #[allow(dead_code)]
```
Comment (jackh726) [inline chalk-solve/src/infer.rs:83]: Yeah, just remove this.

File: chalk-engine/src/logic.rs [hunk 2]
```diff
@@ -328,9 +328,9 @@ impl<C: Context> Forest<C> {
         infer: &mut dyn InferenceTable<C>,
         subgoal: &C::GoalInEnvironment,
     ) -> Option<(C::UCanonicalGoalInEnvironment, C::UniverseMap)> {
-        match infer.truncate_goal(context.interner(), subgoal) {
-            Some(_) => None,
-            None => Some(infer.fully_canonicalize_goal(context.interner(), subgoal)),
+        match infer.goal_needs_truncation(context.interner(), subgoal) {
```
Comment (nikomatsakis) [inline chalk-engine/src/logic.rs:331]: Nit: `if infer.goal_needs_truncation(..) { .. } else { ... }`

File: chalk-engine/src/logic.rs [hunk 3]
```diff
@@ -385,9 +385,9 @@ impl<C: Context> Forest<C> {
         // affect completeness when it comes to subgoal abstraction.
         let inverted_subgoal = infer.invert_goal(context.interner(), subgoal)?;
 
-        match infer.truncate_goal(context.interner(), &inverted_subgoal) {
-            Some(_) => None,
-            None => Some(infer.fully_canonicalize_goal(context.interner(), &inverted_subgoal)),
+        match infer.goal_needs_truncation(context.interner(), &inverted_subgoal) {
```
Comment (nikomatsakis) [inline chalk-engine/src/logic.rs:388]: Nit: if

**Comment by Areredify** [comment]
@nikomatsakis done

**Pull request merged**